### DISCUSSION TITLE
How to limit visibility of a @Factory in test

### DISCUSSION THREAD

**Discussion by nedelva** [micronaut-projects/micronaut-core#10129] [Q&A] [answered]
In one of my tests I would like to replace a  bean with a different (mock) implementation, so that I would end up with this configuration:

> src/main/java/mypackage/MyController
> src/test/java/mypackage/MyControllerRegularTest
> src/test/java/mypackage/MyControllerCornerCaseTest

MyController is using a bean implementing ValidationStrategy interface, like so:
```java
@Controller
public class MyController {
    private final ValidationStrategy validator;
   
    public MyController(@Named("validationStrategyNoOp") ValidationStrategy validationStrategy) { //initializations
    }
    //other stuff
}
```
_ValidationStrategy_ has a default implementation, _ValidationStrategyImpl_:
```java
@Singleton 
public class ValidationStrategyImpl { //the impl omitted for brevity
}
``` 
AND a NO-OP implementation _ValidationStrategyNoOp_ that is temporarily referenced in the controller via the Named annotation.

Now, in the _MyControllerRegularTest_ test  I can see that the implementation _ValidationStrategyNoOp_ is injected in the MyController as expected.

For the _MyControllerCornerCaseTest_ test I need a _different_ implementation of _ValidationStrategy_, so I thought I would solve the problem by providing a bean factory in the test, like this:
```java
@MicronautTest 
public class MyControllerCornerCaseTest {
    
     /* actual tests omitted */ 
    @Factory
    static class StrategyProvider {
          @Bean @Replaces(ValidationStrategyNoOp.class) @Named("validationStrategyNoOp")
          public ValidationStrategy specialCaseStrategy() {
                return new ValidationStrategy() {
                      //some mock implementation useful ONLY for this test!
                }
          }
    } 
}
```
If I run ONLY _MyControllerCornerCaseTest_, the tests pass (I see this mock implementation being injected by the factory).

BUT the problem is that now the _MyControllerRegularTest_  sees too this bean factory and it won't inject anymore the _ValidationStrategyNoOp_ , but the one provided by the _MyControllerCornerCaseTest.StrategyProvider_ factory!
 
So the question is, is there another and better way of specifying an interface implementation, with limited visibility in just one(or a few) test(s)?

**Comment by graemerocher** [comment, marked as answer]
This is what `@MockBean(ValidationStrategyNoOp.class)` is for. It isolates the bean to only the test

**Reply by nedelva** [reply]
Solid advice @graemerocher ! Here's what I ended up with:
```java
@MicronautTest 
public class MyControllerCornerCaseTest {
    
     /* actual tests omitted */ 
    @Factory
    static class StrategyProvider {
          //@Bean @Replaces(ValidationStrategyNoOp.class) 
          @MockBean(ValidationStrategyNoOp.class)
          @Named("validationStrategyNoOp")
          public ValidationStrategy specialCaseStrategy() {
                return new ValidationStrategy() {
                      //some mock implementation useful ONLY for this test!
                }
          }
    } 
}
```
Note that **Named** annotation still is needed, otherwise there will be an ambiguity at the injection point in _MyController_.
